File: server-python/src/revturbine/core/state/storage.py
"""Storage abstraction — Python port of @revt-eng/core/state/storage.ts.

Provides a pluggable interface that works in any Python context.
``InMemoryStorage`` is the default; downstream callers can plug in a
file-backed or DB-backed implementation by satisfying the
``RevTurbineStorage`` protocol.

``JsonFileStorage`` is Python-specific: the TS ``storage.ts`` only adds
a browser ``BrowserStorage`` (``localStorage``/``sessionStorage``),
which is out of scope for the Python port per plan 33 REQ-14
(browser-only). Python local-mode services that gate across process
restarts (gunicorn/uvicorn workers, Celery, cron) need durable
interaction/cap state, so a thin JSON-file backend fills that role.

Source: revturbine-scaffold/src/core/state/storage.ts
"""

from __future__ import annotations

import contextlib
import json
import os
import tempfile
from pathlib import Path
from typing import Protocol, runtime_checkable

__all__ = ["InMemoryStorage", "JsonFileStorage", "RevTurbineStorage"]


@runtime_checkable
class RevTurbineStorage(Protocol):
    """Minimal storage interface mirroring the TS Web Storage API subset.

    Source: storage.ts:17-21
    """

    def get_item(self, key: str) -> str | None: ...
    def set_item(self, key: str, value: str) -> None: ...
    def remove_item(self, key: str) -> None: ...


class InMemoryStorage:
    """Process-local Map-backed storage. Data does not persist beyond the
    current Python process.

    Source: storage.ts:29-43
    """

    def __init__(self) -> None:
        self._store: dict[str, str] = {}

    def get_item(self, key: str) -> str | None:
        return self._store.get(key)

    def set_item(self, key: str, value: str) -> None:
        self._store[key] = value

    def remove_item(self, key: str) -> None:
        self._store.pop(key, None)


class JsonFileStorage:
    """File-backed storage that persists across Python processes.

    The whole key/value map is held in memory and the backing file is a
    single JSON object rewritten atomically on every mutation (write to
    a sibling temp file, then ``os.replace`` — atomic on POSIX and
    Windows). The map is loaded once on construction; a missing,
    unreadable, or malformed file yields an empty map rather than an
    error, mirroring the swallow-on-error resilience of the TS
    ``BrowserStorage`` (``storage.ts:18-45``) — a degraded storage
    backend must never crash decisioning.

    Not a port of any TS symbol: ``BrowserStorage`` is browser-only and
    explicitly out of scope (plan 33 REQ-14). Stdlib only (plan 33
    REQ-12). Single-process-at-a-time durability; concurrent writers on
    the same path are out of scope (use a real KV store for that).

    Source: Python-specific; resilience modeled on storage.ts:18-45.
    """

    def __init__(self, path: str | os.PathLike[str]) -> None:
        self._path = Path(path)
        self._store: dict[str, str] = self._load()

    def _load(self) -> dict[str, str]:
        try:
            raw = self._path.read_text(encoding="utf-8")
        except OSError:
            return {}
        try:
            data = json.loads(raw)
        except (ValueError, TypeError):
            return {}
        if not isinstance(data, dict):
            return {}
        return {str(k): str(v) for k, v in data.items()}

    def _flush(self) -> None:
        try:
            self._path.parent.mkdir(parents=True, exist_ok=True)
            fd, tmp = tempfile.mkstemp(
                prefix=f".{self._path.name}.",
                suffix=".tmp",
                dir=str(self._path.parent),
            )
            try:
                with os.fdopen(fd, "w", encoding="utf-8") as fh:
                    json.dump(self._store, fh, separators=(",", ":"), sort_keys=True)
                os.replace(tmp, self._path)
            except OSError:
                with contextlib.suppress(OSError):
                    os.unlink(tmp)
        except OSError:
            # Swallow — a non-writable path degrades to in-memory-only
            # for this process rather than breaking the caller.
            pass

    def get_item(self, key: str) -> str | None:
        return self._store.get(key)

    def set_item(self, key: str, value: str) -> None:
        self._store[key] = value
        self._flush()

    def remove_item(self, key: str) -> None:
        if self._store.pop(key, None) is not None:
            self._flush()

File: server-rust/src/state/storage.rs
//! Pluggable storage abstraction.
//!
//! [`InMemoryStorage`] is the default. Downstream callers can plug in a
//! file- or DB-backed implementation by satisfying [`RevTurbineStorage`].
//!
//! The browser backends (`localStorage` / `sessionStorage`) are deliberately
//! absent — browser-only, and a REQ-14 non-goal for the headless server SDK.
//!
//! Source: revturbine-scaffold/src/core/state/storage.ts

use std::collections::HashMap;

/// Minimal storage interface, mirroring the Web Storage API subset the TS
/// core uses.
///
/// Source: storage.ts:17-21 (RevTurbineStorage)
pub trait RevTurbineStorage {
    /// Read a value, or `None` when the key is absent.
    fn get_item(&self, key: &str) -> Option<String>;
    /// Write a value, replacing any existing one.
    fn set_item(&mut self, key: &str, value: &str);
    /// Delete a key. Absent keys are not an error.
    fn remove_item(&mut self, key: &str);
}

/// Lets a **borrowed** store satisfy the trait, so a caller can share one
/// backing store across several consumers instead of moving it into the first
/// one. Without this, `CapEnforcer<&mut dyn RevTurbineStorage>` does not
/// compile and every consumer needs its own store — which would silently
/// partition the state they are all supposed to see.
impl<T: RevTurbineStorage + ?Sized> RevTurbineStorage for &mut T {
    fn get_item(&self, key: &str) -> Option<String> {
        (**self).get_item(key)
    }
    fn set_item(&mut self, key: &str, value: &str) {
        (**self).set_item(key, value);
    }
    fn remove_item(&mut self, key: &str) {
        (**self).remove_item(key);
    }
}

/// Process-local storage. Data does not persist beyond the current process.
///
/// This is the only backend the headless SDK ships: `LocalRuntime` is
/// stateless and in-memory by design, and injecting durable storage is a
/// deliberate caller decision rather than a default.
///
/// Source: storage.ts:29-43 (InMemoryStorage)
#[derive(Debug, Default)]
pub struct InMemoryStorage {
    store: HashMap<String, String>,
}

impl InMemoryStorage {
    /// An empty store.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }
}

impl RevTurbineStorage for InMemoryStorage {
    fn get_item(&self, key: &str) -> Option<String> {
        self.store.get(key).cloned()
    }

    fn set_item(&mut self, key: &str, value: &str) {
        self.store.insert(key.to_string(), value.to_string());
    }

    fn remove_item(&mut self, key: &str) {
        self.store.remove(key);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trips_values() {
        let mut s = InMemoryStorage::new();
        assert_eq!(s.get_item("k"), None);
        s.set_item("k", "v");
        assert_eq!(s.get_item("k"), Some("v".to_string()));
        s.set_item("k", "v2");
        assert_eq!(s.get_item("k"), Some("v2".to_string()));
        s.remove_item("k");
        assert_eq!(s.get_item("k"), None);
    }

    #[test]
    fn removing_an_absent_key_is_not_an_error() {
        let mut s = InMemoryStorage::new();
        s.remove_item("never-set");
        assert_eq!(s.get_item("never-set"), None);
    }
}
